fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java

```diff
@@ -624,11 +648,22 @@ private Query getFilterQuery(Set<ApiFilter> filters) {
         } finally {
             lock.readLock().unlock();
         }
-        return new SinglePagePagination<>(
+        page = new SinglePagePagination<>(
                 Collections.unmodifiableList(filteredDimRows.stream().collect(Collectors.toList())),
                 paginationParameters,
                 documentCount
         );
+        if (LUCENE_CACHE_SIZE > 0) {
+            try {
+                SinglePagePagination<DimensionRow> clone = SerializationUtils.clone(page);
+                synchronized (queryCache) {
+                    queryCache.put(new Pair<>(query, paginationParameters), clone);
+                }
```

Comment:
Synchronizing on the cache is evil (although necessary in some scenarios) and is going to impact your overall throughput. Maybe we should be looking at a ConcurrentLinkedHashMap implementation for this (like the one available at https://github.com/ben-manes/caffeine)